File: LeetCodeJS/FindIfPathExistsInGraph_1971.js
class FindIfPathExistsInGraph_1971 {
    validPath(n, edges, source, destination) {
        if (source === destination) return true;

        // Create adjacency list
        const graph = Array.from({ length: n }, () => []);
        for (const [u, v] of edges) {
            graph[u].push(v);
            graph[v].push(u);
        }

        // Use an array to mark visited nodes
        const visited = Array(n).fill(false);
        visited[source] = true;

        return this.dfs(source, destination, graph, visited);
    }

    dfs(node, destination, graph, visited) {
        if (node === destination) return true;

        for (const neighbor of graph[node]) {
            if (!visited[neighbor]) {
                visited[neighbor] = true;
                if (this.dfs(neighbor, destination, graph, visited)) return true;
            }
        }

        return false;
    }
}

module.exports = FindIfPathExistsInGraph_1971;

File: LeetCodeKotlin/FindIfPathExistsInGraph_1971.kt
class FindIfPathExistsInGraph_1971 {
    private fun recursiveDfs(node: Int, destination: Int, graph: Array<MutableList<Int>>, visited: BooleanArray): Boolean {
        if (node == destination) return true
        
        for (neighbor in graph[node]) {
            if (!visited[neighbor]) {
                visited[neighbor] = true
                if (recursiveDfs(neighbor, destination, graph, visited)) return true
            }
        }
        return false
    }

    fun validPath(n: Int, edges: Array<IntArray>, source: Int, destination: Int): Boolean {
        if (source == destination) return true
        
        val graph = Array(n) { mutableListOf<Int>() }
        for (edge in edges) {
            graph[edge[0]].add(edge[1])
            graph[edge[1]].add(edge[0])
        }
        
        val visited = BooleanArray(n)
        visited[source] = true
        
        return recursiveDfs(source, destination, graph, visited)
    }
}